### PR TITLE
Filtering, sorting and tests

### PR DESCRIPTION
**Pull request by lclarksl**
Apologies for the mammoth pull request, but had to change quite a few things to make it testable.

Key Things:
- Created factories for object models that need to be shared between controllers/directives and their tests
- Renamed filterOption to statusCheckbox
- Added ngHTML2JS preprocessor to load views into cache before running tests so that directives don't have to make GET requests to retrieve their views.
- Updated dashboard and sidebar to have filtering and sorting and added tests

### CODE REVIEW

File: tests/unit/dashboard/sidebar/sidebar.spec.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    });
+
+    it("isSelected returns false if status excluded", function() {
+        excludedStatusesList.addStatus(2);
+        var element = compileElement(excludedStatusesList, pref);
+        var isolatedScope = element.isolateScope();
+        expect(isolatedScope.isSelected(2)).toBe(false);
+    });
+
+    it("toggleSelected excludes included state", function() {
```
Comment (jebbinSCL) [inline tests/unit/dashboard/sidebar/sidebar.spec.js]: Test title is a bit confusing. Rename for clarity?

File: tests/unit/dashboard/sidebar/sidebar.spec.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        var element = compileElement(excludedStatusesList, pref);
+        var isolatedScope = element.isolateScope();
+        isolatedScope.toggleSelected(1);
+        expect(isolatedScope.excludedStatusesList.statuses).toEqual([1]);
+    });
+
+    it("toggleSelected includes excluded state", function() {
```
Comment (jebbinSCL) [inline tests/unit/dashboard/sidebar/sidebar.spec.js]: Test title is a bit confusing. Rename for clarity? (similar to above)

File: tests/unit/dashboard/dashboard.controller.spec.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                });
+                return {
+                    getSpeakers: spy
+                }
+            });
+        });
+    });
+
+    beforeEach(inject(function(_$controller_, _$rootScope_, _speakerService_) {
+        $controller = _$controller_;
```
Comment (jebbinSCL) [inline tests/unit/dashboard/dashboard.controller.spec.js]: Looking at code like this in the angular docs, at some point there is a var $controller etc somewhere on the page, but we aren't using that here. Should there be one? Same for $rootScope and mockSpeakerService

Comment (lclarksl) [inline tests/unit/dashboard/dashboard.controller.spec.js]: We use the $controller in the createController() function

Comment (lclarksl) [inline tests/unit/dashboard/dashboard.controller.spec.js]: Sorry, misunderstood you. Have fixed.

**Comment by lclarksl** [comment]
Given directive links their own functions (all hail John Papa) and renamed the tests

**Comment by jebbinSCL** [comment]
:shipit:

**Comment by thullSL** [comment]
:shipit:

**Pull request merged**